core/api/jrpc/value_converter.hpp

```diff
@@ -40,6 +67,32 @@ namespace kagome::api {
     return value;
   }
 
+  inline jsonrpc::Value makeValue(const primitives::Api &val) {
+    using VectorType = jsonrpc::Value::Array;
+    VectorType data;
+
+    data.reserve(2);
+    data.emplace_back(makeValue(val.first));
+    data.emplace_back(makeValue(val.second));
+
+    return std::move(data);
```

Comment:
Why doesn't NRVO work here?

Reply:
Actually you have no guarantees whether the compiler will execute RVO/NRVO optimization or any optimizations. You need to explore in each specific situation. It can avoid skip named return value opt if you have different return objs, different scope of tmp objs and target while handling exceptions, inline asm or even if compiler "thinks" that code is much "complex"(you don't have the explanation of this in Meyers books:)) ). So talking about RVO/NRVO in context of functions that uses non POD-structures is not correct for modern compilers. More over compiler can use "inline", than you don't need to think about nrvo at all, bacause it will have only rval obj with a small scope, so it can construct data in directly in target obj.
This code seems not so simple, because of vector, array several exception types etc. In fact this code will use move semantics even without std::move(), but not NRVO.
Proof uses -O3: https://godbolt.org/z/4NjAiM

Reply:
Okay, if you're sure you know what you're doing in this case, I'll trust you. But I still tend to believe that generally modern compilers are good in optimizing such widespread cases as return values, and doing explicit moves may prevent an optimization, so I'll keep avoiding it, unless writing code for microcontrollers = )
It is still impressive that you apply that deep knowledge of compiler's work even for such plain functions.

Reply:
I mean that it is very dangerous to believe in rvo/nrvo, because if it will not be performed you will have a great performance reduce on expensive copy operations. More over it depends on compiler and it's version. If you want an explicit control of objects creation a reference transfer will be much better ```void some_func(some_type &obj);```. In fact rvo/nrvo uses the same way.